linalg/ginkgo.hpp

```diff
@@ -107,169 +309,334 @@ struct ResidualLogger : gko::log::Logger
+   {
+      return this->executor;
+   };
+
+private:
+   std::shared_ptr<gko::Executor> executor;
+
+};
 
 /**
-* This class forms the base class for all of Ginkgo's iterative solvers.  The
-* various derived classes only take the additional data that is specific to them
-* and solve the given linear system. The entire collection of solvers that
-* Ginkgo implements is available at the Ginkgo documentation and manual pages,
+* This class forms the base class for all of Ginkgo's preconditioners.  The
+* various derived classes only take the additional data that is specific to them.
+* The entire collection of preconditioners that Ginkgo implements is available
+* at the Ginkgo documentation and manual pages,
 * https://ginkgo-project.github.io/ginkgo/doc/develop.
 *
 * @ingroup GinkgoWrappers
 */
-class GinkgoIterativeSolverBase
+class GinkgoPreconditioner : public Solver
 {
 public:
    /**
     * Constructor.
     *
-    * The @p exec_type defines the paradigm where the solution is computed.
-    * It is a string and the choices are "omp" , "reference" or "cuda".
-    * The respective strings create the respective executors as given below.
-    *
-    * Ginkgo currently supports three different executor types:
+    * The @p exec defines the paradigm where the solution is computed.
+    * Ginkgo currently supports four different executor types:
     *
     * +    OmpExecutor specifies that the data should be stored and the
     *      associated operations executed on an OpenMP-supporting device (e.g.
     *      host CPU);
-    * ```
-    * auto omp = gko::create<gko::OmpExecutor>();
-    * ```
     * +    CudaExecutor specifies that the data should be stored and the
     *      operations executed on the NVIDIA GPU accelerator;
-    * ```
-    * if(gko::CudaExecutor::get_num_devices() > 0 ) {
-    *    auto cuda = gko::create<gko::CudaExecutor>();
-    * }
-    * ```
+    * +    HipExecutor specifies that the data should be stored and the
+    *      operations executed on the GPU accelerator using HIP;
     * +    ReferenceExecutor executes a non-optimized reference implementation,
     *      which can be used to debug the library.
-    * ```
-    * auto ref = gko::create<gko::ReferenceExecutor>();
-    * ```
-    *
-    * The following code snippet demonstrates the using of the OpenMP executor
-    * to create a solver which would use the OpenMP paradigm to the solve the
-    * system on the CPU.
+    */
+   GinkgoPreconditioner(GinkgoExecutor &exec);
+
+   /**
+    * Destructor.
+    */
+   virtual ~GinkgoPreconditioner() = default;
+
+   /**
+    * Generate the preconditioner for the given matrix @p op.
+    * Calling this function is only required when creating a
+    * preconditioner for use with another MFEM solver; to use with
+    * a Ginkgo solver, get the LinOpFactory  pointer through @p GetFactory()
+    * and pass to the Ginkgo solver constructor.
+    */
+   virtual void SetOperator(const Operator &op);
```

Comment:
Yes, `op` has to be a `SparseMatrix`, and the implementation should throw an error if the cast to `SparseMatrix` fails. 
You're right that it should be in the documentation rather than waiting for people to run into an error, though.  :)  
I added it in 4862e92cc914e580423d72e9bbc973a4f60008a7